### ISSUE TITLE
HTTP ERROR 431 reauthentication error

### ISSUE THREAD

**Issue by sinist3rr** [deployKF/deployKF#96]
### Checks

- [X] I have searched the [existing issues](https://github.com/deployKF/deployKF).
- [X] This issue is NOT specific to the CLI. (If so, please open an issue on the [CLI repo](https://github.com/deployKF/cli/issues))

### deployKF Version

0.1.2

### Kubernetes Version

```shell
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.17", GitCommit:"22a9682c8fe855c321be75c5faacde343f909b04", GitTreeState:"clean", BuildDate:"2023-08-25T02:13:39Z", GoVersion:"go1.19.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.9+vmware.1", GitCommit:"d1d7c19c9b6265a8dcd1b2ab2620ec0fc7cee784", GitTreeState:"clean", BuildDate:"2022-12-14T06:23:39Z", GoVersion:"go1.18.9", Compiler:"gc", Platform:"linux/amd64"}
```

### Description

I regularly get an HTTP ERROR 431 if I re-authenticate after my Keycloak session has expired.

When re-authenticating I have cookies taking up to 12 KB. But Istio default max request headers allowed is 60 KiB. 
So I don't quite understand where exactly the error is happening.

Headers from web browser (I shortened the cookies for readability):

```
Request URL:
https://kubeflow.example.com/
Request Method:
GET
Status Code:
431 Request Header Fields Too Large
Referrer Policy:
no-referrer

Response Headers:
Date: Mon, 05 Feb 2024 07:10:08 GMT
Server: istio-envoy
X-Envoy-Upstream-Service-Time: 4

Request Headers:
:authority: kubeflow.example.com
:method: GET
:path: /
:scheme: https
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,ru;q=0.8
Cache-Control: max-age=0
Cookie:_deploykf_token_csrf_VKAn1hSV=1o5...tdcU=; _deploykf_token_csrf_nVpidspH=0z-WK...Wbg0=; _deploykf_token_csrf_IR4Ry80n=MeeeDcxLqM...sADFoQ_4=; _deploykf_token_csrf_N1tzT4g2=rJ_IkJ...j2fY=; _deploykf_token_csrf_KjON0Iki=i_Sio-1i8oINTp...EXo=; _deploykf_token_csrf_RY4KCmlu=qKg_U...gDE=; _deploykf_token_csrf_5LkCft-w=inY...OzoG8=; _deploykf_token_csrf_KbluAgpv=aCH2aK...kWfFQ=; _deploykf_token_csrf_T4LJnrLo=bKjg7Q...HOFEtk=; _deploykf_token_csrf_d9r7AanU=fR...J-q0oV7dA=; _deploykf_token_csrf_nL3oqHrj=8lN4p...syokh-v90=; _deploykf_token_csrf_Fa--VPTS=Nao...BXJyhYNY=; _deploykf_token_csrf_34b8tK4Z=LdJo3bbcq1YcL...7LKMEsJuZJLo=; _deploykf_token_csrf_boGY58YM=CCL..DZV0=; _deploykf_token_csrf_hkfYh2Yo=F6z..HgXRQcw=; _deploykf_token_csrf_dzYrnVj3=h_f..sFxqpuZ_QQ=; _deploykf_token_csrf_PhZSUxe9=HP..d4W0=; _deploykf_token_csrf_fXiBlrj0=JJ-1..Dg=; _deploykf_token_csrf_TFZ84rFM=DW..E_qC_s=; _deploykf_token_csrf_0oyI..VrV0=; _deploykf_token_csrf_SxK4cyfs=2U..YcMo=; _deploykf_token_csrf_2M0Bji9d=b-i6..d2c0=; _deploykf_token_csrf_NkpC573h=cc...6vc=; _deploykf_token_csrf_28jWR_Dp=mu...KU=; _deploykf_token_csrf_SOzhamEs=9P...r4=; _deploykf_token_csrf_yZdrbvE_=J0...Rs=; _deploykf_token=0k...kbr4=
Sec-Ch-Ua: "Chromium";v="121", "Not A(Brand";v="99"
Sec-Ch-Ua-Mobile: ?0
Sec-Ch-Ua-Platform: "Linux"
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: same-site
Sec-Fetch-User: ?1
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36
```

### deployKF Values (Optional)

```yaml
## --------------------------------------
  ##             deploykf-auth
  ## --------------------------------------
  deploykf_auth:
    namespace: deploykf-auth

    ## dex configs
    ##
    dex:

      ## dex connectors
      ##
      connectors:
        - type: oidc
          id: keycloak
          name: keycloak
          config:
            issuer: "https://keycloak.example.com/realms/kubeflow"
            clientID: kubeflow
            clientSecret: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
            redirectURI: "https://kubeflow.example.com/dex/callback"
            scopes:
            - openid
            - profile
            - email
            insecureSkipEmailVerified: true
            insecureEnableGroups: true
            userIDKey: email
            userNameKey: email

      ## dex OpenID Connect clients
      ##
      clients:

        ## OpenID client for oauth2-proxy (deployKF Dashboard)
        oauth2Proxy:
          clientId: "oauth2-proxy"
          clientSecret:
            existingSecret: "generated--dex-oauth2-proxy-client"
            existingSecretKey: "client_secret"
            generateSecret: true


    ## oauth2-proxy configs
    ##
    oauth2Proxy:

      ## oauth2-proxy cookie configs
      ##
      cookie:
        secret:
          existingSecret: "generated--oauth2-proxy-cookie-secret"
          existingSecretKey: "cookie_secret"
          generateSecret: true
```

**Comment by thesuperzapper** [comment]
@sinist3rr it looks like you have a LOT of `_deploykf_token_csrf_xxxxxx` cookies, which is not normal. 

I wonder if it's because the dashboard is making lots of requests (in the background) when your session has expired, and you are building up a CSRF token for each one.

Can you confirm (using your browser's "network" panel in its "dev tools") if before you see the 431 status you are being redirected many times or if some other strange thing is happening that might be causing a lot of parallel requests (which are each getting Set-Cookie response headers for CSRF cookies).

**Comment by thesuperzapper** [comment]
For reference, I think that this is primarily caused by the fact that XHR/Ajax requests (like the ones the Notebooks page will make in the background) are sometimes redirected to authentication by oauth2-proxy, and thus get a new CSRF cookie for each background request.

I have raised this issue with oauth2-proxy https://github.com/oauth2-proxy/oauth2-proxy/issues/2481 and I am encouraging them to put a temporary fix in raised by another oauth2-proxy user here https://github.com/oauth2-proxy/oauth2-proxy/issues/2383.

However, it's possible that we can actually implement their proposed workaround in deployKF itself, by adding a Lua-code EnvoyFilter to the oauth2-proxy route that will remove the oldest CSRF cookies, if there are more than X in a request.

The Lua filter could be attach to `envoy_on_request()` and when too many CSRF cookies are in a request, it can do a direct response with a 307 redirect to the same URL path, but telling the browser to remove some of the cookies using a set-cookie header.

**Comment by thesuperzapper** [comment]
Also, I found a situation that can easily cause the ever-increasing number of CSRF cookies in deployKF.

It's when your token is expired, and you have a Kubeflow Notebook open. This is because Jupyter will constantly try to get the Notebook state, but will redirected to auth, and get a new CSRF cookie each time.

**Comment by sinist3rr** [comment]
> @sinist3rr it looks like you have a LOT of `_deploykf_token_csrf_xxxxxx` cookies, which is not normal.
> 
> I wonder if it's because the dashboard is making lots of requests (in the background) when your session has expired, and you are building up a CSRF token for each one.
> 
> Can you confirm (using your browser's "network" panel in its "dev tools") if before you see the 431 status you are being redirected many times or if some other strange thing is happening that might be causing a lot of parallel requests (which are each getting Set-Cookie response headers for CSRF cookies).

Hi. I'm confirming a lot of 302 redirects (sessions, kernels, contents, terminals) that access oauth2-proxy and create these cookies. Especially in the notebook tab.

Also, you really described the most typical flow of my work, when notebook is opened in a separate tab, then after a few hours the token lifetime runs out, I re-authenticate, and get a 431 error and observe a huge number of _deploykf_token_csrf cookies.

**Comment by thesuperzapper** [comment]
@sinist3rr Because of https://github.com/deployKF/deployKF/issues/96, this issue should largely be resolved in the upcoming 0.1.4 release of deployKF.

We also have a new sign-in page which you might want to try (see https://github.com/deployKF/deployKF/pull/100), as that will permanently fix this issue in the long term, but it will not be enabled by default in 0.1.4.

**Comment by sinist3rr** [comment]
> @sinist3rr first, you should not see token expiry very often, because refresh tokens are automatically used in the background, and are valid for 90 days as long as they are used at least every 7 days.
> 
> Can you confirm that refresh tokens work when you add `offline_access` to your scopes? For more information, see the docs [connecting deployKF and Keycloak docs](https://www.deploykf.org/guides/platform/deploykf-authentication/#keycloak). Also, please tell me if there are any errors on that page.
> 

I added offline_access to my scopes.
And it looks like the bug is no longer reproducing. Now it is indeed no longer required to re-login. I think the token refresh works now.
Thanks a lot for this tip!
